iterator.js

```diff
@@ -19,6 +19,10 @@ function Iterator(iterable) {
 
     if (iterable instanceof Iterator) {
         return iterable;
+    } else if (iterable.reduce) {
+        this.reduce = function() {
+            return iterable.reduce.apply(iterable,arguments);
+        }
```

Comment:
The effect is that the collection then have a possibility to determine behaviour of the iterator on itself. Basing on the fact that other operations of the iterator are using reduce operation, the collection determines the whole iterator behaviour rewriting the only (ideally) reduce operation.